<a href="https://colab.research.google.com/github/DanB1421/DATA601/blob/main/Homework_7_Brilliant_690.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 1

In [1]:
import nltk
nltk.download('movie_reviews')
nltk.download('punkt')
from textblob.sentiments import NaiveBayesAnalyzer
from textblob import TextBlob
from os import path

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Unzipping corpora/movie_reviews.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


In [3]:
ObjRead = open("/content/Burbank.txt", "r")
txtContent = ObjRead.read();
ObjRead.close()

In [5]:
blob = TextBlob(txtContent, analyzer=NaiveBayesAnalyzer())
blob.sentiment

Sentiment(classification='pos', p_pos=1.0, p_neg=1.0487590501191368e-20)

The probability that the Burbank text is going to negative is about 0%

# Exercise 2

In [6]:
testimonial = TextBlob(txtContent)
testimonial.sentiment

Sentiment(polarity=0.09869334480780263, subjectivity=0.3790877796901893)

The overall sentiment of the Burbank text is positive, and the text is about 37.9% subjective.

# Exercise 3

In [8]:
import nltk
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
import random
from textblob import Word
from textblob import TextBlob
from os import path

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [9]:
ObjRead = open("/content/Burbank.txt", "r")
txtContent = ObjRead.read();
ObjRead.close()
testimonial = TextBlob(txtContent)

In [10]:
nouns = []
for word, tag in testimonial.tags:
    if tag == 'NN':
        nouns.append(word.lemmatize())

In [12]:
print("This text is about...")
for item in random.sample(nouns, 10):
    word = Word(item)
    print(word)

This text is about...
firm
argue
shift
traffic
wider
issue
northbound
traffic
north
meeting


# Exercise 4

In [24]:
import pandas as pd
import spacy
import numpy as np
import sklearn as sk
import matplotlib.pyplot as plt
import seaborn as sns

In [25]:
data_amazon = pd.read_table('/content/amazon_cells_labelled.txt')
data_imdb = pd.read_table('/content/imdb_labelled.txt')
data_yelp = pd.read_table('/content/yelp_labelled.txt')
combined_col = [data_amazon, data_imdb, data_yelp]

In [26]:
print(data_amazon.columns)

Index(['So there is no way for me to plug it in here in the US unless I go by a converter.', '0'], dtype='object')


In [27]:
for colname in combined_col:
    colname.columns = ["Review","Label"]
for colname in combined_col:
    print(colname.columns)

Index(['Review', 'Label'], dtype='object')
Index(['Review', 'Label'], dtype='object')
Index(['Review', 'Label'], dtype='object')


In [28]:
company = [ "Amazon", "imdb", "yelp"]
comb_data = pd.concat(combined_col,keys = company)
print(comb_data.shape)
comb_data.head()

(2745, 2)


Review  Label
Amazon 0                        Good case, Excellent value.      1
       1                             Great for the jawbone.      1
       2  Tied to charger for conversations lasting more...      0
       3                                  The mic is great.      1
       4  I have to jiggle the plug to get it to line up...      0

In [33]:
comb_data.to_csv("Sentiment_Analysis_Dataset")

In [35]:
print(comb_data.columns)
print(comb_data.isnull().sum())

Index(['Review', 'Label'], dtype='object')
Review    0
Label     0
dtype: int64


In [38]:
import en_core_web_sm
from spacy.lang.en.stop_words import STOP_WORDS
nlp = en_core_web_sm.load()
stopwords = list(STOP_WORDS)
print(stopwords)

['he', 'whom', 'therefore', 'i', 'next', 'also', 'seemed', 'whether', 'whoever', 'nine', 'should', 'itself', 'bottom', '’d', 'for', 'still', "'ll", '‘ll', 'my', 'own', 'her', 'beforehand', 'just', '‘ve', 'around', 'go', 'former', 'rather', 'anyhow', 'a', 'thereafter', 'yours', 'onto', 'anyway', 'down', 'further', 'part', '’re', 'moreover', 'became', 'as', 'will', 'how', 'others', 'behind', 'seem', 'serious', 'forty', '‘s', 'name', 'those', 'thereupon', 'to', 'front', 'everywhere', 'except', 'nevertheless', 'the', 'whither', '’ve', 'six', 'whence', 'nowhere', 'without', 'yourselves', 'never', "'re", 'therein', 'give', 'always', 'yourself', 'eleven', 'else', 'was', 'at', 'seeming', "'s", 'eight', 'us', 'most', 'become', 'yet', 'herein', 'make', 'per', 'whatever', 'mostly', 'n‘t', 'move', 'due', 'done', 'few', 'hereafter', 'being', 'last', 'thence', 'must', 'already', 'hereupon', 'in', 'from', 'regarding', '‘m', 'someone', 'of', 'any', 'himself', 'into', 'wherein', 'out', 'very', 'elsewhe

In [39]:
import string
punctuations = string.punctuation
from spacy.lang.en import English
parser = English()

In [40]:
def my_tokenizer(sentence):
    mytokens = parser(sentence)
    mytokens = [word.lemma_.lower().strip() if word.lemma_ != "-PRON-" else word.lower_]
    mytokens = [word for word in mytokens if word not in stopwords and word not in punctuations]
    return mytokens

In [41]:
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.metrics import accuracy_score
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC

In [42]:
class predictors(TransformerMixin):
    def transform(self, X, **transform_params):
        return [clean_text(text) for text in X]
    def fit(self, X, y, **fit_params):
        return self
    def get_params(self, deep=True):
        return {}

In [43]:
def clean_text(text):
    return text.strip().lower()

In [44]:
from spacy import tokenizer
vectorizer = CountVectorizer(tokenizer = tokenizer, ngram_range=(1,1))
classifier = LinearSVC()

In [45]:
tfvectorizer = TfidfVectorizer(tokenizer = tokenizer)

In [46]:
from sklearn.model_selection import train_test_split
X = comb_data['Review']
ylabels = comb_data['Label']

X_train, X_test, y_train, y_test = train_test_split(X, ylabels, test_size=0.2)

In [47]:
pipe_countvect = Pipeline([("cleaner", predictors()),
                 ('vectorizer', CountVectorizer()),
                 ('classifier', LinearSVC())])

pipe_countvect.fit(X_train,y_train)
sample_prediction = pipe_countvect.predict(X_test)

In [49]:
for (sample,pred) in zip(X_test,sample_prediction):
    print(sample,"Prediction=>",pred)
print("Accuracy for the test dataset: ",pipe_countvect.score(X_test,y_test))
print("Accuracy for the X_test/sample prediction: ",pipe_countvect.score(X_test,sample_prediction))

The update procedure is difficult and cumbersome. Prediction=> 0
I liked the patio and the service was outstanding. Prediction=> 1
Best headset ever!!!. Prediction=> 1
Hell no will I go back Prediction=> 0
I won't say any more - I don't like spoilers, so I don't want to be one, but I believe this film is worth your time.   Prediction=> 1
The scenes with the "oh-so-mature" neighbour-girl are a misplace.   Prediction=> 1
Delicious NYC bagels, good selections of cream cheese, real Lox with capers even. Prediction=> 1
So in a nutshell: 1) The restaraunt smells like a combination of a dirty fish market and a sewer. Prediction=> 0
Great service and food. Prediction=> 1
It was that loud.Glad to say that the Plantronics 510 maintains a flawless connection to my cell and with no static during normal use. Prediction=> 0
The cast is good.   Prediction=> 1
Oh yeah, and the storyline was pathetic too.   Prediction=> 0
I've also had problems with the phone reading the memory card in which I always t

In [50]:
print("Accuracy for the train dataset: ",pipe_countvect.score(X_train,y_train))

Accuracy for the train dataset:  0.9981785063752276
